### PR TITLE
build: enable compilation on win32 and macOS

### PR DESCRIPTION
**Pull request by therealbobo**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area build

> /area engine

> /area tests

> /area proposals

/area CI

**What this PR does / why we need it**:

This enables falco to be build on Windows and MacOS.

**Which issue(s) this PR fixes**:

https://github.com/falcosecurity/falco/issues/2672

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new(build): enable compilation on win32 and macOS.
```

### CODE REVIEW

File: userspace/falco/app/actions/print_page_size.cpp [hunk 1]
```diff
@@ -24,7 +24,15 @@ falco::app::run_result falco::app::actions::print_page_size(falco::app::state& s
 {
 	if(s.options.print_page_size)
 	{
+#ifndef _WIN32
 		long page_size = getpagesize();
+#else
+		SYSTEM_INFO sysInfo;
+
+		GetSystemInfo(&sysInfo);
```
Comment (LucaGuerra) [inline userspace/falco/app/actions/print_page_size.cpp:32]: Looks like there is no header file included for this function :thinking:

File: userspace/falco/app/actions/print_support.cpp [hunk 2]
```diff
@@ -37,23 +39,61 @@ falco::app::run_result falco::app::actions::print_support(falco::app::state& s)
...
 		support["engine_info"] = infos.as_json();
 
+#ifndef _WIN32
 		support["system_info"]["sysname"] = sysinfo.sysname;
 		support["system_info"]["nodename"] = sysinfo.nodename;
 		support["system_info"]["release"] = sysinfo.release;
 		support["system_info"]["version"] = sysinfo.version;
 		support["system_info"]["machine"] = sysinfo.machine;
+#else
+		support["system_info"]["sysname"] = "Windows";
```
Comment (LucaGuerra) [inline userspace/falco/app/actions/print_support.cpp:75]: Wouldn't it be a little cleaner to create a function that is able to output the `system_info`, have one for linux/mac and one for Windows so we can better separate the two implementations?

File: userspace/falco/app/actions/create_requested_paths.cpp [hunk 3]
```diff
@@ -93,23 +94,15 @@ falco::app::run_result falco::app::actions::create_requested_paths(falco::app::s
...
+
+    std::filesystem::path dirPath(path);
+
+    try {
+        std::filesystem::create_directories(dirPath);
+    } catch (const std::exception& ex) {
+		return -1;
+    }
+
+    return 0;
```
Comment (LucaGuerra) [inline userspace/falco/app/actions/create_requested_paths.cpp:106]: <3

File: userspace/falco/stats_writer.cpp [hunk 4]
```diff
@@ -48,7 +54,10 @@ static void timer_handler(int signum)
 
 bool stats_writer::init_ticker(uint32_t interval_msec, std::string &err)
 {
+#if !defined(_WIN32)
```
Comment (LucaGuerra) [inline userspace/falco/stats_writer.cpp:57]: This is getting a bit hard to follow

File: userspace/falco/logger.h [hunk 5]
```diff
@@ -18,7 +18,18 @@ limitations under the License.
...
+#ifdef _WIN32
+#define	LOG_EMERG	0
+#define	LOG_ALERT	1
+#define	LOG_CRIT	2
+#define	LOG_ERR		3
+#define	LOG_WARNING	4
+#define	LOG_NOTICE	5
+#define	LOG_INFO	6
+#define	LOG_DEBUG	7
+#else
```
Comment (LucaGuerra) [inline userspace/falco/logger.h:30]: can we make an enum for `falco_logger::level` instead?

File: userspace/falco/app/actions/print_page_size.cpp (hunk 1 above)
Comment (LucaGuerra) [reply userspace/falco/app/actions/print_page_size.cpp:32]: I have seen this in other files as well, a bit confused as to where this is imported from

**Comment by LucaGuerra** [comment]
Thanks a lot @therealbobo !! This is very useful to make Falco more cross-platform compatible <3 

My main issues here is mostly about readability because sometimes it's easy to get lost in a bunch of `ifdef`s. Would also love to hear @FedeDP 's opinion about this as he usually has good ideas :)

File: userspace/falco/event_drops.h [hunk 6]
```diff
@@ -29,7 +29,7 @@ limitations under the License.
 // detecting a syscall event drop.
 enum class syscall_evt_drop_action : uint8_t
 {
-	IGNORE = 0,
+	DISREGARD = 0,
```
Comment (LucaGuerra) [inline userspace/falco/event_drops.h:32]: It's not really clear to me why we need this. Is it because just having anything that is called `IGNORE` (even if it's a scoped enum) ends up being a compilation error? Seems a little weird but not impossible!

File: userspace/falco/app/actions/print_support.cpp (hunk 2 above)
Comment (therealbobo) [reply userspace/falco/app/actions/print_support.cpp:75]: That makes sense! 👍

File: userspace/falco/event_drops.h (hunk 6 above)
Comment (therealbobo) [reply userspace/falco/event_drops.h:32]: The problem is this define: `winbase.h:479:#define IGNORE 0` :/

File: userspace/falco/app/actions/print_support.cpp [hunk 7]
```diff
@@ -32,28 +38,84 @@ static std::string read_file(const std::string &filename)
...
+	support["system_info"]["nodename"] = sysinfo.nodename;
+	support["system_info"]["release"] = sysinfo.release;
+	support["system_info"]["version"] = sysinfo.version;
+	support["system_info"]["machine"] = sysinfo.machine;
+	return 0;
+}
+#endif
+
+#ifdef _WIN32
+static int get_win32_sysinfo(nlohmann::json &support)
```
Comment (jasondellaluce) [inline userspace/falco/app/actions/print_support.cpp:60]: Why not call both of these with the same name? That way we don't need any ifdef when invoking them.

File: .github/workflows/reusable_build_packages.yaml [hunk 8]
```diff
@@ -216,3 +216,63 @@ jobs:
...
+
+      - name: Run unit Tests
+        run: |
+          build/unit_tests/Release/falco_unit_tests.exe
+
+      - name: Upload Falco win32 package
+        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        with:
+          name: falco-${{ inputs.version }}-win32.exe
+          path: |
```
Comment (jasondellaluce) [inline .github/workflows/reusable_build_packages.yaml:247]: Question -- can't we update cpack options and just use `make package` and grab the package generated by cmake?

Comment (FedeDP) [reply .github/workflows/reusable_build_packages.yaml:253]: Agree with Jason!

File: .github/workflows/reusable_build_packages.yaml [hunk 9]
```diff
@@ -216,3 +216,69 @@ jobs:
           name: falco-${{ inputs.version }}-wasm.tar.gz
           path: |
             ${{ github.workspace }}/build/falco-${{ inputs.version }}-wasm.tar.gz
+
+  build-win32-package:
```
Comment (FedeDP) [inline .github/workflows/reusable_build_packages.yaml:220]: Can't we use a matrix `build-others-packages` like we do eg: on libs?

File: CMakeLists.txt [hunk 10]
```diff
@@ -109,30 +113,55 @@ endif()
...
-if(NOT EMSCRIPTEN)
+if(LINUX)
   set(FALCO_SECURITY_FLAGS "${FALCO_SECURITY_FLAGS} -Wl,-z,relro,-z,now -fstack-protector-strong")
 endif()
-if(CMAKE_BUILD_TYPE STREQUAL "release")
-  set(FALCO_SECURITY_FLAGS "${FALCO_SECURITY_FLAGS} -D_FORTIFY_SOURCE=2")
-endif()
 
-set(CMAKE_COMMON_FLAGS "${FALCO_SECURITY_FLAGS} -Wall -ggdb ${FALCO_EXTRA_FEATURE_FLAGS} ${MINIMAL_BUILD_FLAGS} ${MUSL_FLAGS}")
+if(NOT WIN32)
```
Comment (FedeDP) [inline CMakeLists.txt:120]: Can we move all of this in a `CompilerFlags.cmake` like we do for example in libs?

File: cmake/modules/falcosecurity-libs.cmake [hunk 11]
```diff
@@ -28,15 +28,15 @@ else()
   # In case you want to test against a fork of falcosecurity/libs just pass the variable -
   # ie., `cmake -DFALCOSECURITY_LIBS_REPO=<your-gh-handle>/libs ..` 
   if (NOT FALCOSECURITY_LIBS_REPO)
-    set(FALCOSECURITY_LIBS_REPO "falcosecurity/libs")
+    set(FALCOSECURITY_LIBS_REPO "therealbobo/libs")
```
Comment (FedeDP) [inline cmake/modules/falcosecurity-libs.cmake:31]: I assume you are just testing things; in case please add back the `wip` status :)

File: unit_tests/falco/test_configuration.cpp [hunk 12]
```diff
@@ -109,7 +109,11 @@ TEST(Configuration, configuration_environment_variables)
     std::string env_var_value = "envVarValue";
     std::string env_var_name = "ENV_VAR";
     std::string default_value = "default";
+#ifndef _WIN32
     setenv(env_var_name.c_str(), env_var_value.c_str(), 1);
+#else
+    _putenv_s(env_var_name.c_str(), env_var_value.c_str());
```
Comment (FedeDP) [inline unit_tests/falco/test_configuration.cpp:115]: We could use a macro on top of this file perhaps?

File: userspace/falco/outputs_program.cpp
Comment (FedeDP) [inline userspace/falco/outputs_program.cpp:1]: Since we now won't support all outputs on all OSes, we could build just the list of supported outputs on each OS from CMakeLists. Would this be feasible?
Then we only need a change on falco_outputs.cpp to only include the supported outputs.

File: userspace/falco/stats_writer.cpp (hunk 4 above)
Comment (FedeDP) [reply userspace/falco/stats_writer.cpp:57]: I would just have different versions of init_ticker for each OS.

File: cmake/modules/falcosecurity-libs.cmake (hunk 11 above)
Comment (therealbobo) [reply cmake/modules/falcosecurity-libs.cmake:31]: Yeap! Exactly! Just for testing purposes :)

File: userspace/falco/app/actions/print_page_size.cpp (hunk 1 above)
Comment (FedeDP) [reply userspace/falco/app/actions/print_page_size.cpp:35]: Cc @therealbobo

**Review by jasondellaluce — APPROVED** [review]
/approve

**Comment by FedeDP** [comment]
Updated the release note label to allow the PR to get merged.
